Fixes and refactors death examine code

Gamer025 opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request
Fixes https://github.com/tgstation/tgstation/issues/55890
Examine death bodies should now display the correct text
Changed the logic so that the message tells you if the players is:

- Still in his body (`[t_He] [t_is] limp and unresponsive; there are no signs of life...`) or
- A ghost that could enter the body again (`[t_He] [t_is] limp and unresponsive; there are no signs of life and [t_his] soul has departed, but the link is not yet fully broken...`) or
- No ghost, that can reenter the body and no key (`[t_He] [t_is] limp and unresponsive; there are no signs of life and [t_his] soul has lost the will to live...`)

Also refactored the code a bit:

- Moved the death examine message generation to a proc so that you can work with returns
- Removed pushed_do_not_resuscitate since its not needed this way

While looking at the code I also realized that we have a var `can_reenter_corpse` for ghosts.
This var gets set to false when you press DNR for example.
However when pressing DNR your mind of your ghost also gets nulled so that already prevents you from reentering your body.
Maybe this var could be removed too.

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game
Fixes examine text for death players
<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Changelog
:cl: Gamer025
fix: Fixed the examine text for death players (death players not appearing as dead)
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

Gamer025 committed:
Fixes carbon examine

Fixes death examine for carbons and some refactors

Gamer025 committed:
Collapse if

Gamer025 committed:
Split if

Split if to make it more easy to read

Jared-Fogle commented:
oh wait youre actually using a real gpg key nevermind

Cyberboss commented:
> We give the verified label only to dubiously verified committers
Absolute stat of /tg/

LemonInTheDark commented on code/modules/mob/living/carbon/carbon.dm:
```diff
@@ -11,6 +11,15 @@
 	if(!mapload)  //I don't want no gas leaks on my space ruin you hear?
 		RegisterSignal(src, COMSIG_LIVING_DEATH, .proc/attach_rot)
 
+	//this is very slightly better than it was because you can use it more places. still can't do \his[src] though.
+	t_He = p_they(TRUE)
+	t_His = p_their(TRUE)
+	t_his = p_their()
+	t_him = p_them()
+	t_has = p_have()
+	t_is = p_are()
+	t_es = p_es()
+
```
Couldn't this change during gameplay? At least in theory?

Cyberboss replied:
Yeah, this gave me pause on this PR. Seems to be a very fragile cache and not worth it.

Gamer025 replied:
Yeah I now realize why this is done this way (as cache) and just added the necessary vars to my proc instead of moving the cache

LemonInTheDark commented on code/modules/mob/living/carbon/human/human_helpers.dm:
```diff
@@ -220,3 +220,16 @@
+///Returns death message for mob examine text
+/mob/living/carbon/human/proc/generate_death_examine_text()
+	var/mob/dead/observer/ghost = get_ghost(TRUE, TRUE)
+	if(getorgan(/obj/item/organ/brain) && !key) //The death mob has a brain and no client/player that is assigned to the mob
```
Is key set to false when a player dies? Does converting from client to key maintain behavior?

Gamer025 replied:
I was looking at the code before https://github.com/tgstation/tgstation/pull/55399 was merged and before that we used
`if(getorgan(/obj/item/organ/brain) && !key && !get_ghost(FALSE, TRUE))` to determine the "soul has departed" part of the message. Because of this I changed it back to key.
As far as I know key is used internally by Byond to know which mob should be controlled by a player when he joins the server. When you die and start to spectate as ghost the key var gets nulled because you are now controlling the ghost. This means if someone died, left the server but did not spectate the if will still be true and it will display "[t_He] [t_is] limp and unresponsive; there are no signs of life..." even though the player has disconnected. (after all the player could reconnect)
Client as far as I know determines if someone is currently controlling the mob. So if I would check for client it would jump between "[t_He] [t_is] limp and unresponsive; there are no signs of life and [t_his] soul has lost the will to live..." (there is no ghost if you never leave your body) and "[t_He] [t_is] limp and unresponsive; there are no signs of life..." depending if the play is currently disconnected or not.

Im honestly sure how many infos we want to give players about the status of someone else so I changed the messages to the format I described in the PR description.

Gamer025 committed:
Undo pronouns cache movement

Yeah moving the cache to carbon init was probably a bad idea
Adds cache to helpers proc

Gamer025 committed:
Undo obscure_name at human level

Gamer025 committed:
Wrong pronoun var

Gamer025 committed:
Merge if again ...

This actually needs to be one condition, since this will behave wrong if someone was to change just the can_reenter_corpse and keep the mind linked to a ghost.

Rohesie commented on code/modules/mob/living/carbon/human/human_helpers.dm:
```diff
@@ -220,3 +220,19 @@
+	var/t_He = p_they(TRUE)
+	var/t_his = p_their()
+	var/t_is = p_are()
+	if(getorgan(/obj/item/organ/brain) && !key) //The death mob has a brain and no client/player that is assigned to the mob
+		if (!ghost || !ghost.can_reenter_corpse)  //And there is no ghost that could reenter the body
+			//There is no way this mob can in any normal way get a player, so he lost the will to live
+			return "<span class='deadsay'>[t_He] [t_is] limp and unresponsive; there are no signs of life and [t_his] soul has lost the will to live...</span>"
+		else //There is a ghost and it can reenter
+			//This mob has a ghost linked that could still reenter the body, so the soul only departed
+			return "<span class='deadsay'>[t_He] [t_is] limp and unresponsive; there are no signs of life and [t_his] soul has departed...</span>"
+	//If no special cases matches return the default death message
+	return "<span class='deadsay'>[t_He] [t_is] limp and unresponsive; there are no signs of life...</span>"
```
```suggestion
	if(!getorgan(/obj/item/organ/brain) || key)
		return "<span class='deadsay'>[t_He] [t_is] limp and unresponsive; there are no signs of life...</span>" //Default death message
	//The death mob has a brain and no client/player that is assigned to the mob
	if(!ghost?.can_reenter_corpse)  //And there is no ghost that could reenter the body
		//There is no way this mob can in any normal way get a player, so they lost the will to live
		return "<span class='deadsay'>[t_He] [t_is] limp and unresponsive; there are no signs of life and [t_his] soul has lost the will to live...</span>"
	//This mob has a ghost linked that could still reenter the body, so the soul only departed
	return "<span class='deadsay'>[t_He] [t_is] limp and unresponsive; there are no signs of life and [t_his] soul has departed...</span>"
```

Gamer025 replied:
Won't this change the logic when there is no ghost?
Like for example someone spectated (so the dead mob now has no key) and joined again via some ghost spawn.
`!ghost?.can_reenter_corpse` would now be null so false and it would reach the last return instead of the "lost hiss will to live" return which should normally be triggered since there is nobody who can reenter the body.
I guess the easiest would be to change line 233 to `isnull(ghost) || !ghost.can_reenter_corpse`?

Rohesie replied:
`!(null?.can_reenter_corpse)` -> `!(null)` -> `TRUE`
The `?.` operator gets resolved before the `!` negation one, and doesn't skip it.

If you wanted the behavior you describe you'd have to write it as `if(ghost && !ghost.can_reenter_corpse)`

Gamer025 committed:
Early returns

Co-authored-by: Rohesie <rohesie@gmail.com>

Rohesie commented on code/modules/mob/living/carbon/human/human_helpers.dm:
```diff
@@ -220,3 +220,18 @@
+		//There is no way this mob can in any normal way get a player, so they lost the will to live
+		return "<span class='deadsay'>[t_He] [t_is] limp and unresponsive; there are no signs of life and [t_his] soul has lost the will to live...</span>"
+	//This mob has a ghost linked that could still reenter the body, so the soul only departed
+	return "<span class='deadsay'>[t_He] [t_is] limp and unresponsive; there are no signs of life and [t_his] soul has departed...</span>"
```
```suggestion
	return "<span class='deadsay'>[t_He] [t_is] limp and unresponsive; there are no signs of life and [t_his] soul has departed, but the link is not yet fully broken...</span>"
```
People are used to equating "soul has departed" with DNR. It would be useful to clarify here.

Gamer025 replied:
That's actually a really good idea for letting people know about the change
I was thinking about how to best explain in this in the changelog maybe but I really like this idea

Gamer025 committed:
Add clarification to soul has departed message

Co-authored-by: Rohesie <rohesie@gmail.com>
